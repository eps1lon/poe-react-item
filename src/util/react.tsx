import * as React from 'react';

// text nodes dont need keys
export const CommaSeparator = () => ', ';

export const LineBreak = (key: string) => <br key={key} />;

export function intersperse(
  nodes: React.ReactNode[],
  renderSeparator: (key: string) => React.ReactNode = CommaSeparator,
) {
  if (nodes.length === 0) {
    return [];
  }

  return nodes.slice(1).reduce(
    (joined: React.ReactNode[], element, index) => {
      // push instead of concat's returnval because concat is expensive
      joined.push(renderSeparator(`separator-${index}`), element);
      return joined;
    },
    [nodes[0]],
  );
}
