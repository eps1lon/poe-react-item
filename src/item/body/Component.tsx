import { PureComponent } from 'react';

import { ItemProps } from '../poe';

export interface Props {
  item: ItemProps;
}

export default class Body extends PureComponent<Props> {
  render() {
    return 'Body';
  }
}
