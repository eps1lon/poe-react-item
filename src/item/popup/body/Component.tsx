import * as React from 'react';

import { Item } from '../../poe';
import { Mod } from '../../../mod/poe';
import { Stat } from '../../../stat/poe';
import Stats from '../../../stat/Stats';
import { Intersperse } from '../../../util/react';
import Separator from '../Separator';
import Properties from './Properties';
import Requirements from './Requirements';

export interface Props {
  item: Item;
  translations: {};
}

export default class Body extends React.PureComponent<Props> {
  public render() {
    const { item, translations } = this.props;
    const { implicits = [], explicits = [], requirements = {} } = item;
    const implicit_stats = this.collectStats(implicits);
    const explicit_stats = this.collectStats(explicits);

    return (
      <section className="body">
        <Intersperse renderSeparator={separator}>
          {Properties.hasAny(item) && (
            <Properties key="properties" properties={item} />
          )}
          {Requirements.hasAny(requirements) && (
            <Requirements key="requirements" requirements={requirements} />
          )}
          {Stats.hasAny(implicit_stats) && (
            <Stats
              key="implicits"
              classname="implicit"
              stats={implicit_stats}
              translations={translations}
            />
          )}
          {Stats.hasAny(explicit_stats) && (
            <Stats
              key="explicits"
              classname="explicit"
              stats={explicit_stats}
              translations={translations}
            />
          )}
        </Intersperse>
      </section>
    );
  }

  private collectStats(mods: Mod[]): Stat[] {
    return mods.reduce(
      (flattened, { stats = [] }) => {
        flattened.push(...stats);
        return flattened;
      },
      [] as Stat[],
    );
  }
}

function separator(key: string) {
  return <Separator key={key} />;
}
