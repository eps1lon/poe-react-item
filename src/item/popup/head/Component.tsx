import classnames from 'classnames';
import * as React from 'react';

import { Item, Rarity } from '../../poe';

import NameLine, { hasNameLine } from './NameLine';
import TypeLine from './TypeLine';

export interface Props {
  item: Item;
}

export default class Head extends React.PureComponent<Props> {
  render() {
    const { item } = this.props;
    const single_line = !hasNameLine(item);

    return (
      <header
        className={classnames({
          'double-line': !single_line,
          shaper: Boolean(item.shaper),
          elder: Boolean(item.elder),
        })}
      >
        <span className="name-left" />
        <NameLine item={item} />
        <TypeLine item={item} />
        <span className="name-right" />
      </header>
    );
  }
}
