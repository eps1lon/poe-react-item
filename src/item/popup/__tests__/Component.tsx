import * as React from 'react';
import * as renderer from 'react-test-renderer';

import Component, { Rarity } from '../ItemPopup';

it('should display Mind Brow Hubris Circlet', () => {
  expect(
    renderer
      .create(
        <Component
          item={{
            base: {
              name: 'Hubris Circlet',
            },
            name: 'Mind Brow',
            rarity: Rarity.rare,
            energy_shield: {
              value: 200,
              augmented: true,
            },
            quality: 6,
            shaper: true,
            requirements: {
              level: {
                value: 69,
              },
              intelligence: {
                value: 154,
              },
            },
            explicitStats: [
              'Place an additional Mine',
              'Socketed Gems are Supported by Level 18 Remote Mine',
              '16% increased Effect of Chill',
              'Socketed Gems are Supported by Level 18 Hypothermia',
              '+1 to Level of Socketed Minion Gems',
              '+67 to maximum Life',
              '18% increased Stun and Block Recovery',
              'Adds 5 to 89 Lightning Damage to Spells',
            ],
          }}
        />,
      )
      .toJSON(),
  ).toMatchSnapshot();
});
