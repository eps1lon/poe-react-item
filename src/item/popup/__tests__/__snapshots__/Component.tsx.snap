// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display Mind Brow Hubris Circlet 1`] = `
<div
  className="poe-item rare"
>
  <header
    className="double-line shaper"
  >
    <span
      className="name-left"
    />
    <div
      className="item-name"
    >
      Mind Brow
    </div>
    <div
      className="item-name type-line"
    >
      <span>
        Hubris Circlet
      </span>
    </div>
    <span
      className="name-right"
    />
  </header>
  <section
    className="body"
  >
    <div
      className="display-property"
    >
      <span>
        quality
      </span>
      : 
      <span
        className="value augmented"
      >
        6%
      </span>
    </div>
    <div
      className="display-property"
    >
      <span>
        energy shield
      </span>
      : 
      <span
        className="value augmented"
      >
        200
      </span>
    </div>
    <div
      className="separator"
    />
    <div
      className="requirements"
    >
      Requires 
      <span>
        level
      </span>
       
      <span
        className="value default"
      >
        69
      </span>
      , 
      <span
        className="value default"
      >
        154
      </span>
       
      <span>
        int
      </span>
    </div>
    <div
      className="separator"
    />
    <div
      className="explicit"
    >
      Place an additional Mine
    </div>
    <div
      className="explicit"
    >
      Socketed Gems are Supported by Level 18 Remote Mine
    </div>
    <div
      className="explicit"
    >
      16% increased Effect of Chill
    </div>
    <div
      className="explicit"
    >
      Socketed Gems are Supported by Level 18 Hypothermia
    </div>
    <div
      className="explicit"
    >
      +1 to Level of Socketed Minion Gems
    </div>
    <div
      className="explicit"
    >
      +67 to maximum Life
    </div>
    <div
      className="explicit"
    >
      18% increased Stun and Block Recovery
    </div>
    <div
      className="explicit"
    >
      Adds 5 to 89 Lightning Damage to Spells
    </div>
  </section>
</div>
`;
